### ISSUE TITLE
PostgreSQL inserts very slow.

### ISSUE THREAD

**Issue by jamessewell** [ShadowTraffic/requests#2]
Running from MacOS Docker to a local PostgreSQL instance (on the host node) I'm getting less than 1 insert per second.

When I run with `--stdout` speed is as expected.

My config file is as follows:

```
{
  "generators": [
    {
      "table": "sensors",
      "vars": {
        "id": {
          "_gen": "oneOf",
          "choices": {
            "_gen": "stringSeries",
            "bounds": [
              1,
              100
            ],
            "expr": "sensor-~d"
          }
        }
      },
      "row": {
        "sensorid": {
          "_gen": "var",
          "var": "id"
        },
        "ts": {
          "_gen": "weightedOneOf",
          "choices": [
            {
              "weight": 98,
              "value": {
                "_gen": "formatDateTime",
                "ms": {
                  "_gen": "now"
                }
              }
            },
            {
              "weight": 2,
              "value": {
                "_gen": "formatDateTime",
                "ms": {
                  "_gen": "subtract",
                  "args": [
                    {
                      "_gen": "now"
                    },
                    {
                      "_gen": "uniformDistribution",
                      "bounds": [
                        604800,
                        604800000
                      ]
                    }
                  ]
                }
              }
            }
          ]
        }
      },
      "stateMachine": {
        "_gen": "stateMachine",
        "for": {
          "_gen": "var",
          "var": "id"
        },
        "initial": "start",
        "transitions": {
          "start": "update",
          "update": "update"
        },
        "states": {
          "start": {
            "row": {
              "value": {
                "_gen": "normalDistribution",
                "mean": 50,
                "sd": 5
              }
            }
          },
          "update": {
            "row": {
              "value": {
                "_gen": "add",
                "args": [
                  {
                    "_gen": "uniformDistribution",
                    "bounds": [
                      -10,
                      100
                    ]
                  },
                  {
                    "_gen": "previousEvent",
                    "path": [
                      "row",
                      "value"
                    ]
                  }
                ]
              }
            }
          }
        }
      },
      "localConfigs": {
        "throttle": {
          "ms": {
            "_gen": "uniformDistribution",
            "bounds": [
              0,
              100
            ]
          }
        }
      }
    }
  ],
  "connections": {
    "pg": {
      "kind": "postgres",
      "connectionConfigs": {
        "host": "host.docker.internal",
        "port": 28815,
        "username": "james",
        "password": "",
        "db": "postgres"
      }
    }
  }
}

**Comment by jamessewell** [comment]
Huh looks like this was a docker on Mac strangeness thing.

Having said that .... it looks like you're opening and closing connections for each INSERT. This is very under-performant, can you keep it open?

**Comment by MichaelDrogalis** [comment]
Investigating now!

> Huh looks like this was a docker on Mac strangeness thing.

Can you comment on what that was? I'm also on a Mac and can't reproduce that.

**Comment by MichaelDrogalis** [comment]
I've reproduced the connection opening and closing on every write. It's odd because `.close` isn't being explicitly called in that path, but I'm tracking it down. Thanks for raising!

**Comment by MichaelDrogalis** [comment]
Fixed and released in `0.0.32`—silly error on my part. Thanks for helping me track it down! 🫡

I'm now seeing about 250 rows/s inserted. Still not fast, but avoiding any really dumb methods of inserting.

Logs while inserting look good now:

```
2023-11-19 21:48:05.593 UTC [128] LOG:  connection received: host=192.168.65.1 port=22572
2023-11-19 21:48:05.604 UTC [128] LOG:  connection authenticated: identity="postgres" method=scram-sha-256 (/var/lib/postgresql/data/pg_hba.conf:128)
2023-11-19 21:48:05.604 UTC [128] LOG:  connection authorized: user=postgres database=mydb
2023-11-19 21:48:05.607 UTC [128] LOG:  execute <unnamed>: SET extra_float_digits = 3
2023-11-19 21:48:05.607 UTC [128] LOG:  execute <unnamed>: SET application_name = 'PostgreSQL JDBC Driver'
2023-11-19 21:48:05.610 UTC [128] LOG:  execute <unnamed>: SELECT EXISTS (SELECT * FROM information_schema.tables WHERE table_name='sensors' AND table_catalog='mydb')
2023-11-19 21:48:05.615 UTC [128] LOG:  execute <unnamed>: INSERT INTO sensors (sensorId, timestamp, value) VALUES ($1, $2, $3)
	RETURNING *
2023-11-19 21:48:05.615 UTC [128] DETAIL:  parameters: $1 = 'sensor-78', $2 = '2023-11-19 13:48:05.614-08', $3 = '37.87004012137835'
2023-11-19 21:48:05.619 UTC [128] LOG:  execute <unnamed>: SELECT c.oid, a.attnum, a.attname, c.relname, n.nspname, a.attnotnull OR (t.typtype = 'd' AND t.typnotnull), a.attidentity != '' OR pg_catalog.pg_get_expr(d.adbin, d.adrelid) LIKE '%nextval(%' FROM pg_catalog.pg_class c JOIN pg_catalog.pg_namespace n ON (c.relnamespace = n.oid) JOIN pg_catalog.pg_attribute a ON (c.oid = a.attrelid) JOIN pg_catalog.pg_type t ON (a.atttypid = t.oid) LEFT JOIN pg_catalog.pg_attrdef d ON (d.adrelid = a.attrelid AND d.adnum = a.attnum) JOIN (SELECT 16469 AS oid , 1 AS attnum UNION ALL SELECT 16469, 2 UNION ALL SELECT 16469, 3) vals ON (c.oid = vals.oid AND a.attnum = vals.attnum)
2023-11-19 21:48:05.623 UTC [128] LOG:  execute <unnamed>: INSERT INTO sensors (sensorId, timestamp, value) VALUES ($1, $2, $3)
	RETURNING *
2023-11-19 21:48:05.623 UTC [128] DETAIL:  parameters: $1 = 'sensor-36', $2 = '2023-11-19 13:48:05.622-08', $3 = '67.90001859271145'
2023-11-19 21:48:05.628 UTC [128] LOG:  execute <unnamed>: INSERT INTO sensors (sensorId, timestamp, value) VALUES ($1, $2, $3)
	RETURNING *
2023-11-19 21:48:05.628 UTC [128] DETAIL:  parameters: $1 = 'sensor-75', $2 = '2023-11-19 13:48:05.627-08', $3 = '30.945234403420557'
2023-11-19 21:48:05.632 UTC [128] LOG:  execute <unnamed>: INSERT INTO sensors (sensorId, timestamp, value) VALUES ($1, $2, $3)
	RETURNING *
2023-11-19 21:48:05.632 UTC [128] DETAIL:  parameters: $1 = 'sensor-43', $2 = '2023-11-19 13:48:05.631-08', $3 = '44.981380241574826'
2023-11-19 21:48:05.636 UTC [128] LOG:  execute S_1: INSERT INTO sensors (sensorId, timestamp, value) VALUES ($1, $2, $3)
	RETURNING *
2023-11-19 21:48:05.636 UTC [128] DETAIL:  parameters: $1 = 'sensor-16', $2 = '2023-11-19 13:48:05.635-08', $3 = '36.43770583390375'
2023-11-19 21:48:05.637 UTC [128] LOG:  disconnection: session time: 0:00:00.044 user=postgres database=mydb host=192.168.65.1 port=22572
```

**Comment by MichaelDrogalis** [comment]
(Feel free to reopen this or open a new issue to discuss further Postgres perf improvements since there's clearly a lot of room to make it better)
